nixos/modules/services/hardware/pid-fan-controller.nix

```diff
@@ -0,0 +1,182 @@
+{
+  lib,
+  config,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.pid-fan-controller;
+  heatSource = {
+    options = {
+      name = lib.mkOption {
+        type = lib.types.uniq lib.types.nonEmptyStr;
+        description = "name of heat source";
+      };
+      wildcard_path = lib.mkOption {
+        type = lib.types.nonEmptyStr;
+        description = ''wildcard path of heat source hwmon "temp_input", can contain wildcards'';
+      };
+      PID_params = {
+        set_point = lib.mkOption {
+          type = lib.types.ints.unsigned;
+          description = "set point of controller in °C";
+        };
+        P = lib.mkOption {
+          description = "K_p of PID controller";
+          type = lib.types.float;
+        };
+        I = lib.mkOption {
+          description = "K_i of PID controller";
+          type = lib.types.float;
+        };
+        D = lib.mkOption {
+          description = "K_d of PID controller";
+          type = lib.types.float;
+        };
+      };
+    };
+  };
+
+  fan = {
+    options = {
+      wildcard_path = lib.mkOption {
+        type = lib.types.str;
+        description = ''
+          wildcard path of the hwmon "pwm" file.
+          If the fans are not to be found in /sys/class/hwmon/hwmon* the corresponding
+          kernel module (like "nct6775") needs to be added to `boot.kernelModules`.
+          See: [hwmon](https://www.kernel.org/doc/html/latest/hwmon/index.html)
+        '';
+      };
+      min_pwm = lib.mkOption {
+        default = 0;
+        type = lib.types.ints.between 0 255;
+        description = "minimum PWM fan speed";
+      };
+      max_pwm = lib.mkOption {
+        default = 255;
+        type = lib.types.ints.between 0 255;
+        description = "maximum PWM fan speed";
+      };
+      cutoff = lib.mkOption {
+        default = false;
+        type = lib.types.bool;
+        description = "whether to stop fan when `min_pwm` is reached";
+      };
+      heat_pressure_srcs = lib.mkOption {
+        type = lib.types.nonEmptyListOf lib.types.str;
+        description = "heat pressure sources which are affected by the fan";
+      };
+    };
+  };
+in
+{
+  options.services.pid-fan-controller = {
+    enable = lib.mkEnableOption "Enable PID fan controller";
+    package = lib.mkPackageOption pkgs "pid-fan-controller" { };
+    settings = {
+      interval = lib.mkOption {
+        default = 500;
+        type = lib.types.int;
+        description = "Interval between controller cycles.";
+      };
+      heat_srcs = lib.mkOption {
+        type = lib.types.listOf (lib.types.submodule heatSource);
+        description = "list of heat sources";
+        example = ''
```

Comment:
The example is likely a big to big to properly render on search.nixos.org.

Reply:
I just halved the length of the example. Is there some example of the limits search.nixos.org is able to display?

Reply:
wdym? Yeah, it's a bit big, but `nixos-search` and the manual should be fine with it.

Reply:
It is not really a hard limit but if things get to big, they generally don't look great in the layout. Should be fine now though.